src/main/java/org/craftercms/engine/util/servlet/ConfigAwareProxyServlet.java

```diff
@@ -0,0 +1,158 @@
+import org.slf4j.LoggerFactory;
+import org.springframework.context.ApplicationContext;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.commons.lang3.StringUtils.removeStart;
+
+/**
+ * Extension of {@link ProxyServlet} that uses the current site configuration
+ *
+ * @author joseross
+ * @since 3.1.6
+ */
+public class ConfigAwareProxyServlet extends ProxyServlet implements AppContextAwareServlet {
+
+    private static final Logger logger = LoggerFactory.getLogger(ConfigAwareProxyServlet.class);
+
+    public static final String PARAM_CONFIG_PATH = "configPath";
+
+    public static final String CONFIG_KEY_SERVERS = "servers.server";
+
+    public static final String CONFIG_KEY_PATTERNS = "patterns.pattern";
+
+    public static final String CONFIG_KEY_ID = "id";
+
+    public static final String CONFIG_KEY_URL = "url";
+
+    protected ApplicationContext applicationContext;
+
+    protected ContentStoreService contentStoreService;
+
+    protected String configPath;
+
+    protected boolean preview;
+
+    @Override
+    public void setApplicationContext(ApplicationContext applicationContext) {
+        this.applicationContext = applicationContext;
+    }
+
+    @Override
+    public void init() throws ServletException {
+        super.init();
+
+        configPath = getInitParameter(PARAM_CONFIG_PATH);
+
+        contentStoreService = applicationContext.getBean(ContentStoreService.class);
+    }
+
+    @Override
+    protected void initTarget() throws ServletException {
+        // Do nothing ... the target url will be resolved for each request
+    }
+
+    @Override
+    protected void service(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException {
+        String requestUri = request.getRequestURI();
+        logger.debug("Starting execution of proxy request for {}", requestUri);
+
+        SiteContext siteContext = SiteContext.getCurrent();
+
+        if (siteContext == null) {
+            throw new IllegalStateException("Can't resolve site context for current request");
+        }
+
+        // get the site id from the queryString
+        String siteName = siteContext.getSiteName();
+
+        logger.debug("Resolved site {} for proxy request {}", siteName, requestUri);
+
+        // get the target url from the site config
+        String targetUrl = getTargetUrl(siteContext, requestUri);
+
+        logger.debug("Resolved target url {} for proxy request {}", targetUrl, requestUri);
+
+        if (request.getRequestURL().toString().contains(targetUrl)) {
+            logger.debug("Resolved target url for request {} is local, will skip proxy", requestUri);
+            throw new HttpProxyException("Local target url detected");
```

Comment:
I think it might be better to use another exception specific to this case, something like `LocalTargetUrlProxyException`.